src/main/java/org/spongepowered/common/entity/player/SpongeUserData.java

```diff
@@ -417,34 +382,28 @@ public void markDirty() {
         if (this.isConstructing) {
             return;
         }
-        if (!this.isInitialized()) {
-            SpongeCommon.logger()
-                    .warn("Unable to mark user data for [{}] as dirty, data is not initialized! Any changes may be lost.",
-                            this.profile.getId());
-        } else {
-            SpongeUser.dirtyUsers.add(this);
-        }
+        ((SpongeServer) SpongeCommon.server()).userManager().markDirty(this);
     }
 
     public void save() {
-        Preconditions.checkState(this.isInitialized(), "User {} is not initialized", this.profile.getId());
-
-        final LevelStorageSource.LevelStorageAccess storageSource = ((MinecraftServerAccessor) Sponge.server()).accessor$storageSource();
-        final File file = storageSource.getLevelPath(LevelResource.PLAYER_DATA_DIR).resolve(this.uniqueId() + ".dat").toFile();
-        CompoundTag compound;
-        try {
-            compound = NbtIo.readCompressed(new FileInputStream(file));
-        } catch (final IOException ignored) {
-            // Nevermind
-            compound = new CompoundTag();
-        }
-        this.writeCompound(compound);
-        try (final FileOutputStream out = new FileOutputStream(file)) {
-            NbtIo.writeCompressed(compound, out);
-            SpongeUser.dirtyUsers.remove(this);
-            this.invalidate();
-        } catch (final IOException e) {
-            SpongeCommon.logger().warn("Failed to save user file [{}]!", file, e);
+        synchronized (this) {
+            final SpongeUserManager userManager = ((SpongeServer) SpongeCommon.server()).userManager();
+            final LevelStorageSource.LevelStorageAccess storageSource = ((MinecraftServerAccessor) Sponge.server()).accessor$storageSource();
+            final File file = storageSource.getLevelPath(LevelResource.PLAYER_DATA_DIR).resolve(this.uniqueId() + ".dat").toFile();
+            CompoundTag compound;
+            try {
+                compound = NbtIo.readCompressed(new FileInputStream(file));
+            } catch (final IOException ignored) {
+                // Nevermind
```

Comment:
We need to print this as it typically means the compression metadata in the file failed or a file lock. We should not continue at all, if Vanilla is they are playing dangerously...

Reply:
Well, this is actually a bit of a weird one because we've loaded the compound already, but on save we're loading it again and basically ignoring any changes to what we already have - this is the case with data and maybe inventories. So the correct approach is to just use what we have stored and write to that, I think.

We may want to add a way to invalidate any cache on disc in the user manager, but the use case for that will be few and far between, I'm sure.

Reply:
Ah you are performing a merge...I could see that being dangerous because who gets to say exactly what gets merged? Our code or what is on disk? What happens if a cached User is saved right before the hour mark is over yet a player has been playing the last 45 or so mins and the file is saved? I wonder if there is a way to "invalidate" User references but allow API consumers to know this happened? Trying to cover as many scenarios as possible.

Reply:
It's difficult to say, however part of the point of this object is that any access will reuiqre a hit on the cache. That will then reset the timer.

That said, the biggest problem with the previous system was that we were holding data for a `User` and for a `Player` at the same time. This is no longer the case - if a player logs in, the `SpongeUserData` is saved, discarded, and then the `SpongeUserView` switches over to use the `ServerPlayer` as the data source automatically (though the opposite won't happen). We don't need to worry about keeping `User` and `ServerPlayer` in sync any more.

I'm not entirely sure how we could notify users that a `User` has expired, but I don't think it's going to be a significant problem.